### ISSUE TITLE
Skills that are not always active have no modifiers

### ISSUE THREAD

**Issue by stocky37** [LartTyler/MHWDB-API#50]
I believe skills that have modifiers that are not always active (for instance, affinity sliding or heroics) should contain their rank modifiers. It should be up to a consumer of the api to figure out whether or not those modifiers should actually be used (i.e. if simulating or assuming a particular skill is active).

Is there any particular reason that this is the case?

**Comment by LartTyler** [comment]
I left them empty for exactly that reason, because they're not always active. I've been meaning to change that, but I just haven't had a chance to. Ideally, I'd prefer not to add them in there without a way to indicate that they're inactive until certain conditions are met, but I'm really not sure what the best way to present that would be. Any thoughts?

**Comment by stocky37** [comment]
That's a very good question. I'm not entirely sure.

The options I can think of are:

**a)** Don't worry about optional modifiers at all, leave it up to consumers
**b)** Implement a basic "optional" boolean as part of the modifier.
```json
"modifiers": {
  "attack": {
    "value": 5,
    "optional": true
  }
}
```
**c)** Specify *when* a particular value should apply
```json
"modifiers": {
  "attack": {
    "value": 5,
    "applies": "enemyAgitated"
  }
}
```

I think I'm more partial to Option A, but Option B could give you more flexibility in terms of modifiers. For instance you could better support modifiers that are a modifer + bonus, like the element boosts. Option C should probably be avoided because I don't think there would be a nice way to do that.

**Comment by LartTyler** [comment]
From the standpoint of adding as much info as possible, I _really_ like C, but you're right, I'm not sure of a decent way to implement that. Maybe in a later version.

As far as your `optional` field in option B, I like that, but because of the way the modifiers are stored, doing so would break searching for that field.

Which leaves option A, which I'm not a huge fan of since I feel like it creates a bad experience for devs using the API. That means that every person that uses the API for something like what you're doing has to do the legwork of determining which skills should apply their modifiers, and which shouldn't. So here's my thought:

```json
"modifiers": {
  "attack": 5,
},
"modifierInfo": {
  "attack": {
    "optional": true
  }
}
```

That would achieve the goal of providing a way to inform devs that a field shouldn't apply at all times, and would make it easy to do something like add an `applies` field in a future version, so the API can indicate when a modifier should apply. It's not perfect, because simply knowing a value is optional doesn't help determine WHEN it should apply, but at least it gives some indication that the modifiers shouldn't be rolled into any base stat calculations.

If that sounds like it'd work for your needs, I'll plan on adding that in the release after next (since I'm not sure I'll be able to get to it by this weekend).

**Comment by stocky37** [comment]
That definitely works if you want to keep it backwards compatible. I get where you're coming from in terms of Option C, but as a consumer you would still have to implement custom logic to handle each of the different types of "applies", so it seems a bit 6 of one, half a dozen of other to me. With Option B I'll still have to write the custom logic anyway so the modifierInfo won't really benefit me too much, but I'm sure people with other use cases could find it useful.

**Comment by LartTyler** [comment]
I guess my reasoning was that if the `applies` field existed in the API, all you'd have to do as a consumer of the API is add logic for the values of applies. Otherwise, you still have to do that, AND you need to maintain a list of which skills apply under which circumstances.

Either way, the first step is going to be adding the modifiers for skills that aren't always on. Once that's in the API, I'll revisit the issue and decide how I want to mark the skills as active under certain conditions.

**Comment by bolte-17** [comment]
One or two additional suggestions: I think `conditional` might be a better word than `optional` in this context, and I actually like the "applies" idea. I could see some value in being able to get a list of conditions (or even a list of conditions used by the character's current skills) and, for example, rendering a list of checkboxes for the user to toggle on/off. Or even sliders for the user to set what % of time the skill will be active for full-hunt dps estimation.

**Comment by LartTyler** [comment]
Yeah, that's kinda what I was thinking. For the time being, I'm gonna plan on adding the conditionally active skills to the list of modifiers, as if they were just normal modifiers (like what @stocky37 was originally asking for). Once they're in there, I'm gonna revisit this and decide on how I wanna implement the conditional / "applies" idea.

**Comment by stocky37** [comment]
Sounds like a plan!

**Comment by DeveloperPaul123** [comment]
Just to chime in on this, I would say that the modifiers for skills should be independent of the conditions that need to be met for that skill modifier to activate. The approach that makes the most sense to me is to add a `condition` array field that will hold a list of all the conditions that need to be met for the skill to activate. We can also add a `duration` field to indicate how long a modifier is up once it's activated. 

Example:

````json
{
  "id": 107,
  "name": "Affinity Sliding",
  "description": "Sliding increases your affinity for a short time.",
  "conditions" : [
    {
      "activation" : "Slide for 1 or more seconds.",
      "duration" : 15
    }
  ],
  "ranks": [
    {
      "id": 304,
      "skill": 107,
      "level": 1,
      "modifiers": {
        "affinity": 30
      },
      "description": "+30% Affinity for 15 seconds after sliding.",
      "skillName": "Affinity Sliding"
    }
  ]
}
````

**Comment by LartTyler** [comment]
I like that idea. Only change I think I'd make is that the `activation` field should be an enumerated value instead of something free-form. Otherwise we'd need translations for each one, and handling translations on something like that would be harder to implement. Ideally, values could be things like `slide`, `enraged`, `blighted`, etc. Might need to be a longer list to include all possible conditions, but if it's a finite, list it can just be documented and consumers can handle their own localization needs.

**Comment by DeveloperPaul123** [comment]
Good point, though that list will be pretty long I think, but it is still finite so definitely doable.
